cmd/microcloud/ask.go

```diff
@@ -1420,3 +1423,227 @@ func (c *initConfig) askClustered(s *service.Handler, expectedServices []types.S
+				err := json.Unmarshal(bytes, &session)
+				if err != nil {
+					logger.Error("Failed to read join intent", logger.Ctx{"err": err})
+					break
+				}
+
+				// Skip systems which aren't listed in the preseed.
+				if !shared.ValueInSlice(session.Intent.Name, expectedSystems) {
+					continue
+				}
+
+				joinIntents[session.Intent.Name] = session.Intent
+				if len(joinIntents) == len(expectedSystems) {
+					renderCancel()
+				}
+
+			case <-renderCtx.Done():
+				return
+			}
+		}
+	}
+
+	var systems []types.SessionJoinPost
+	if !c.autoSetup {
+		go renderIntentsInteractive()
+
+		// Wait until the table got rendered.
+		// This is important otherwise the table might not be selectable
+		// as it's being built in a go routine.
+		select {
+		case <-rendered:
+		case <-gw.Context().Done():
+			return nil, fmt.Errorf("Failed to render join intents: %w", context.Cause(gw.Context()))
+		}
+
+		var answers []string
+		retry := false
+		err := c.askRetry("Retry selecting systems?", func() error {
+			defer func() {
+				retry = true
+			}()
+
+			fmt.Println("Select which systems you want to join:")
+
+			if retry {
+				err := table.Render(table.rows)
+				if err != nil {
+					return fmt.Errorf("Failed to render table: %w", err)
+				}
+			}
+
+			var err error
+			answers, err = table.GetSelections()
+			if err != nil {
+				return fmt.Errorf("Failed to get join intent selections: %w", err)
+			}
+
+			if len(answers) == 0 {
+				return fmt.Errorf("No system selected")
+			}
+
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		for _, answer := range answers {
+			name := table.SelectionValue(answer, "NAME")
+			for intentName, intent := range joinIntents {
+				if intentName == name {
+					systems = append(systems, intent)
+				}
+			}
+		}
+	} else {
+		go renderIntents()
+
+		select {
+		case <-time.After(c.lookupTimeout):
+		case <-renderCtx.Done():
+		}
+
+		for _, name := range expectedSystems {
+			_, ok := joinIntents[name]
+			if !ok {
+				return nil, fmt.Errorf("System %q hasn't reached out", name)
+			}
+		}
+
+		for _, intent := range joinIntents {
+			systems = append(systems, intent)
+		}
+	}
+
+	return systems, nil
+}
+
+func (c *initConfig) askJoinConfirmation(gw *cloudClient.WebsocketGateway) error {
+	session := types.Session{}
+	err := gw.ReceiveWithContext(gw.Context(), &session)
+	if err != nil {
+		return fmt.Errorf("Failed to read join confirmation: %w", err)
+	}
+
+	if !c.autoSetup {
+		fmt.Printf("\n Received confirmation from system %q\n\n", session.Intent.Name)
+		fmt.Println("Do not exit out to keep the session alive.")
+		fmt.Printf("Complete the remaining configuration on %q ...\n", session.Intent.Name)
+	}
+
+	err = gw.ReceiveWithContext(gw.Context(), &session)
+	if err != nil {
+		return fmt.Errorf("Failed waiting during join: %w", err)
+	}
+
+	if session.Error != "" {
+		return fmt.Errorf("Failed to join system: %s", session.Error)
+	}
+
+	fmt.Println("Successfully joined the cluster")
```

Comment:
This prints at the wrong time, or it needs to change its message.

Currently it prints around when the initiator is bootstrapped, but well before the joiner has joined the cluster.

Reply:
It's still a rather arbitrary distinction. I think the session should either wait until the node has joined all cluster members (since there would be dangling join tokens), or it should end with `Commencing cluster join. Session closed`. or something.

Reply:
I am not following on the dangling tokens though. In which situation would you see this happening?

I meant in case of error. We do try to clean up, though, so it's not a big issue.

What I'm most concerned with is somebody seeing "oh the joiner says it's done" and then try to interact with MicroCloud while the session is still active. This might be a particular problem for preseed because there might be other code that begins to execute as soon as the joiner stops blocking. 

That's why I think the command should be blocking until the initiator reports back to the joiners that the cluster formation is actually fully complete.

Reply:
The preseed blocking is a good point. Even though you can potentially run the operation also on the initiator to work around this (in case it's not specific to this member).

This feels closely related with the topic we have discussed about letting the joiner know in case the initiator observed an error during forming of the cluster.
I'll log this as a separate issue so we can address all of this together in one shot either using a new endpoint `session/exit` or by having a websocket between the systems.

For know I have changed the exit message to the following:

```
Successfully joined the MicroCloud cluster and closing the session.
Commencing cluster join of the remaining services (LXD, MicroCeph, MicroOVN)
```

It's using the local systems list of services and filters out `MicroCloud` as this cluster is already formed.